File: Leetcode_重刷紀錄/0025.Problems735. Asteroid Collision/Asteroid Collision.cpp
class Solution {
public:
    vector<int> asteroidCollision(vector<int>& asteroids) {
        vector<int> stack;  // 使用vector來模擬棧
        for (int ast : asteroids) {
            bool isAlive = true;
            // 只有當當前小行星向左移動，且棧頂小行星向右移動時，才需要處理碰撞
            while (!stack.empty() && ast < 0 && stack.back() > 0) {
                if (stack.back() < -ast) {  // 如果棧頂小行星小於當前小行星的絕對值
                    stack.pop_back();  // 棧頂小行星被摧毀
                    continue;
                } else if (stack.back() == -ast) {  // 如果兩個小行星大小相同
                    stack.pop_back();  // 兩者都被摧毀
                }
                isAlive = false;  // 當前小行星被摧毀
                break;
            }
            if (isAlive) {  // 如果當前小行星存活（未發生碰撞或成功摧毀所有遇到的小行星）
                stack.push_back(ast);  // 將其加入棧中
            }
        }
        return stack;        
    }
};

File: Leetcode_重刷紀錄/0025.Problems735. Asteroid Collision/Asteroid Collision.java
class Solution {
    public int[] asteroidCollision(int[] asteroids) {
        Stack<Integer> stack = new Stack<>();
       
        for (int ast : asteroids) {
            collision: {
                while (!stack.isEmpty() && ast < 0 && stack.peek() > 0) {
                    if (stack.peek() < -ast) {
                        stack.pop();
                        continue;
                    } else if (stack.peek() == -ast) {
                        stack.pop();
                    }
                    // 如果碰撞發生，且棧頂小行星被摧毀或兩者大小相同，則不將目前小行星加入堆疊
                    break collision;
                }
                // 如果沒有發生碰撞，或堆疊為空，或棧頂小行星向左移動，則將目前小行星加入堆疊
                stack.push(ast);
            }
        }

        // 將堆疊中的元素轉換為數組
        int[] result = new int[stack.size()];
        for (int i = result.length - 1; i >= 0; i--) {
            result[i] = stack.pop();
        }
        return result;
    }
}